adapters/criteo/criteo.go

```diff
@@ -13,102 +13,66 @@ import (
+	return []*adapters.RequestData{requestData}, nil
 }
 
-func getCriteoRequestHeaders(criteoRequest *criteoRequest) http.Header {
-	headers := http.Header{}
-
-	// criteoRequest is known not to be nil
-	// If there was an error generating it from newCriteoRequest, the errors will be returned immediately
-	// and this method won't be called
-
-	if criteoRequest.User.CookieID != "" {
-		headers.Add("Cookie", "uid="+criteoRequest.User.CookieID)
-	}
-
-	if criteoRequest.User.IP != "" {
-		headers.Add("X-Forwarded-For", criteoRequest.User.IP)
-	}
-
-	if criteoRequest.User.IPv6 != "" {
-		headers.Add("X-Forwarded-For", criteoRequest.User.IPv6)
-	}
-
-	if criteoRequest.User.UA != "" {
-		headers.Add("User-Agent", criteoRequest.User.UA)
-	}
-
-	return headers
-}
-
-func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
-	if response.StatusCode == http.StatusNoContent {
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
 		return nil, nil
 	}
 
-	if response.StatusCode != http.StatusOK {
-		return nil, []error{&errortypes.BadServerResponse{
-			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
-		}}
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Run with request.debug = 1 for more info.",
+		}
+		return nil, []error{err}
 	}
 
-	bidResponse, err := newCriteoResponseFromBytes(response.Body)
-	if err != nil {
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
 		return nil, []error{err}
 	}
 
-	bidderResponse := adapters.NewBidderResponse()
-	bidderResponse.Bids = make([]*adapters.TypedBid, len(bidResponse.Slots))
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
 
-	for i := 0; i < len(bidResponse.Slots); i++ {
-		bidderResponse.Bids[i] = &adapters.TypedBid{
-			Bid: &openrtb2.Bid{
-				ID:    bidResponse.Slots[i].ArbitrageID,
-				ImpID: bidResponse.Slots[i].ImpID,
-				Price: bidResponse.Slots[i].CPM,
-				AdM:   bidResponse.Slots[i].Creative,
-				W:     bidResponse.Slots[i].Width,
-				H:     bidResponse.Slots[i].Height,
-				CrID:  bidResponse.Slots[i].CreativeCode,
-			},
-			BidType: openrtb_ext.BidTypeBanner,
+	bidResponse := adapters.NewBidderResponse()
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for _, bid := range seatBid.Bid {
+			b := &adapters.TypedBid{
+				Bid:     &bid,
```

Comment:
Please consider using index of `seatBid.Bid` instead of `&bid`. When we use range, the scope of the iteration variable is for that for loop only. This variable is reassigned in every iteration, so when you use address of the variable, you will end up pointing to the last element of the `seatBid.Bid` slice. This can be handled as follow:-

```
for _, seatBid := range response.SeatBid {
	for i := range seatBid.Bid {
		b := &adapters.TypedBid{
			Bid:     &seatBid.Bid[i],
			BidType: openrtb_ext.BidTypeBanner,
		}
		bidResponse.Bids = append(bidResponse.Bids, b)
	}
}
```

Reply:
Hi, thanks for the review. I pushed a fix with the code you proposed.